src/System.Net.Security/src/System/Net/SecureProtocols/_SslStream.cs

```diff
@@ -0,0 +1,709 @@
+
+        private int _InternalOffset;
+        private int _InternalBufferCount;
+
+        private FixedSizeReader _Reader;
+
+        internal _SslStream(SslState sslState)
+        {
+            if (PinnableBufferCacheEventSource.Log.IsEnabled())
+            {
+                PinnableBufferCacheEventSource.Log.DebugMessage1("CTOR: In System.Net._SslStream.SslStream", this.GetHashCode());
+            }
+            _SslState = sslState;
+            _Reader = new FixedSizeReader(_SslState.InnerStream);
+        }
+
+        // If we have a read buffer from the pinnable cache, return it.
+        private void FreeReadBuffer()
+        {
+            if (_InternalBufferFromPinnableCache)
+            {
+                s_PinnableReadBufferCache.FreeBuffer(_InternalBuffer);
+                _InternalBufferFromPinnableCache = false;
+            }
+            _InternalBuffer = null;
+        }
+
+        ~_SslStream()
+        {
+            if (_InternalBufferFromPinnableCache)
+            {
+                if (PinnableBufferCacheEventSource.Log.IsEnabled())
+                {
+                    PinnableBufferCacheEventSource.Log.DebugMessage2("DTOR: In System.Net._SslStream.~SslStream Freeing Read Buffer", this.GetHashCode(), PinnableBufferCacheEventSource.AddressOfByteArray(_InternalBuffer));
+                }
+                FreeReadBuffer();
+            }
+            if (_PinnableOutputBuffer != null)
+            {
+                if (PinnableBufferCacheEventSource.Log.IsEnabled())
+                {
+                    PinnableBufferCacheEventSource.Log.DebugMessage2("DTOR: In System.Net._SslStream.~SslStream Freeing Write Buffer", this.GetHashCode(), PinnableBufferCacheEventSource.AddressOfByteArray(_PinnableOutputBuffer));
+                }
+                s_PinnableWriteBufferCache.FreeBuffer(_PinnableOutputBuffer);
+            }
+        }
+
+        internal int Read(byte[] buffer, int offset, int count)
+        {
+            return ProcessRead(buffer, offset, count, null);
+        }
+
+        internal void Write(byte[] buffer, int offset, int count)
+        {
+            ProcessWrite(buffer, offset, count, null);
+        }
+
+        internal bool DataAvailable
+        {
+            get { return InternalBufferCount != 0; }
+        }
+
+        private byte[] InternalBuffer
+        {
+            get
+            {
+                return _InternalBuffer;
+            }
+        }
+
+        private int InternalOffset
+        {
+            get
+            {
+                return _InternalOffset;
+            }
+        }
+
+        private int InternalBufferCount
+        {
+            get
+            {
+                return _InternalBufferCount;
+            }
+        }
+
+        private void DecrementInternalBufferCount(int decrCount)
+        {
+            _InternalOffset += decrCount;
```

Comment:
You are correct: it's not very clear from the code. It looks like the `InternalBufferCount` is the remainder of the buffer and `InternalOffset` is a pointer within `InternalBuffer`. When the remainder is 0 the offset is set to the end of the buffer.

Instead of comments (that as we've already seen in this PR go out of sync), I'd prefer UTs to pin down the behavior of this class.